File: Services/ScreenshotService.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Drawing.Imaging;
using System.IO;
using System.Linq;
using Crossword.Objects;
using Crossword.ViewModel;
using Point = System.Drawing.Point;
using Brushes = System.Windows.Media.Brushes;

namespace Crossword.Services;

public class ScreenshotService : IScreenshotService
{
    private readonly IDialogService _dialogService;

    public ScreenshotService(IDialogService dialogService)
    {
        _dialogService = dialogService;
    }

    public void CreateCrosswordFiles(List<Word> listWordsGrid, List<Dictionary> listDictionaries, IEnumerable<CellViewModel> cells)
    {
        try
        {
            var topMaxX = 99;
            var leftMaxY = 99;
            var downMaxX = 0;
            var rightMaxY = 0;
            const float sizeCell = 37.938105f;
            SearchMaxCoordinates(ref topMaxX, ref downMaxX, ref leftMaxY, ref rightMaxY, cells);
            if (downMaxX < topMaxX || rightMaxY < leftMaxY)
            {
                _dialogService.ShowMessage("Не найдено ячеек для создания скриншота. Убедитесь, что кроссворд сгенерирован.");
                return;
            }

            var width = (int)((downMaxX - topMaxX + 1) * sizeCell);
            var height = (int)((rightMaxY - leftMaxY + 1) * sizeCell);
            using var img = new Bitmap(width, height);
            img.SetResolution(300, 300);
            using var graphics = Graphics.FromImage(img);
            var listDefinitionRight = new List<string>();
            var listDefinitionDown = new List<string>();
            CreateEmptyGrid(img, graphics, topMaxX, downMaxX, leftMaxY, rightMaxY, sizeCell, listDefinitionRight, listDefinitionDown, cells, listWordsGrid);
            CreateFillGrid(img, graphics, topMaxX, downMaxX, leftMaxY, rightMaxY, sizeCell, cells);
            CreateAnswerFile(listDefinitionRight, listDefinitionDown);
            CreateDefinitionFile(listDefinitionRight, listDefinitionDown, listDictionaries);
            _dialogService.ShowMessage("Кросворд сохранён");
        }
        catch (Exception e)
        {
            _dialogService.ShowMessage($"Ошибка при создании скриншота:\n{e.Message}");
        }
    }

    private void SearchMaxCoordinates(ref int topMaxX, ref int downMaxX, ref int leftMaxY, ref int rightMaxY, IEnumerable<CellViewModel> cells)
    {
        var transparentCells = cells.Where(c => c.Background == Brushes.Transparent).ToList();
        if (!transparentCells.Any())
        {
            topMaxX = 1;
            downMaxX = 0;
            return;
        }

        topMaxX = transparentCells.Min(c => c.X);
        leftMaxY = transparentCells.Min(c => c.Y);
        downMaxX = transparentCells.Max(c => c.X);
        rightMaxY = transparentCells.Max(c => c.Y);
    }

    private void CreateEmptyGrid(Bitmap img, Graphics graphics, int topMaxX, int downMaxX, int leftMaxY, int rightMaxY, float sizeCell, List<string> listDefinitionRight, List<string> listDefinitionDown, IEnumerable<CellViewModel> cells, List<Word> listWordsGrid)
    {
        var blackBrush = new SolidBrush(Color.Black);
        var whiteBrush = new SolidBrush(Color.White);
        using var blackPen = new Pen(Color.Black, 1);
        using var font = new Font("Arial", 4);
        graphics.Clear(Color.White);

        foreach (var cellVm in cells.Where(c => c.X >= topMaxX && c.X <= downMaxX && c.Y >= leftMaxY && c.Y <= rightMaxY))
        {
            if (cellVm.Background == Brushes.Transparent)
            {
                graphics.FillRectangle(whiteBrush, (cellVm.X - topMaxX) * sizeCell, (cellVm.Y - leftMaxY) * sizeCell, sizeCell, sizeCell);
                graphics.DrawRectangle(blackPen, (cellVm.X - topMaxX) * sizeCell, (cellVm.Y - leftMaxY) * sizeCell, sizeCell, sizeCell);
            }
            else
            {
                graphics.FillRectangle(blackBrush, (cellVm.X - topMaxX) * sizeCell, (cellVm.Y - leftMaxY) * sizeCell, sizeCell, sizeCell);
            }
        }

        var numberedStartCells = new Dictionary<Point, int>();
        var numberCounter = 1;
        Cell GetStartingCell(Word word) => word.Cells.FirstOrDefault();
        var orderedWords = listWordsGrid
            .Select(w => new { WordObject = w, StartCell = GetStartingCell(w) })
            .Where(x => x.StartCell != null)
            .OrderBy(x => x.StartCell.Y)
            .ThenBy(x => x.StartCell.X)
            .ThenBy(x => x.WordObject.Right ? 0 : 1)
            .Select(x => x.WordObject);

        foreach (var word in orderedWords)
        {
            var startingCellState = GetStartingCell(word);
            if (startingCellState == null) continue;
            var cellPoint = new Point(startingCellState.X, startingCellState.Y);
            if (!numberedStartCells.TryGetValue(cellPoint, out var wordNumber))
            {
                wordNumber = numberCounter++;
                numberedStartCells[cellPoint] = wordNumber;
                var drawX = (startingCellState.X - topMaxX) * sizeCell;
                var drawY = (startingCellState.Y - leftMaxY) * sizeCell;
                graphics.DrawString(wordNumber.ToString(), font, blackBrush, drawX + 1, drawY + 1);
            }

            var text = wordNumber + ";" + word.WordString;
            if (word.Right)
            {
                listDefinitionRight.Add(text);
            }
            else
            {
                listDefinitionDown.Add(text);
            }
        }

        img.Save("EmptyGrid.png", ImageFormat.Png);
    }

    private void CreateFillGrid(Bitmap img, Graphics graphics, int topMaxX, int downMaxX, int leftMaxY, int rightMaxY, float sizeCell, IEnumerable<CellViewModel> cells)
    {
        var blackBrush = new SolidBrush(Color.Black);
        using var blackPen = new Pen(Color.Black, 1);
        using var font = new Font("Arial", 7);
        graphics.Clear(Color.White);

        foreach (var cell in cells.Where(c => c.X >= topMaxX && c.X <= downMaxX && c.Y >= leftMaxY && c.Y <= rightMaxY))
        {
            var drawX = (cell.X - topMaxX) * sizeCell;
            var drawY = (cell.Y - leftMaxY) * sizeCell;

            if (cell.Background == Brushes.Black)
            {
                graphics.FillRectangle(blackBrush, drawX, drawY, sizeCell, sizeCell);
            }
            else
            {
                var format = new StringFormat
                {
                    LineAlignment = StringAlignment.Center,
                    Alignment = StringAlignment.Center
                };
                graphics.FillRectangle(System.Drawing.Brushes.White, drawX, drawY, sizeCell, sizeCell);
                graphics.DrawRectangle(blackPen, drawX, drawY, sizeCell, sizeCell);
                graphics.DrawString(cell.Content?.ToUpper(), font, blackBrush, drawX + (sizeCell / 2), drawY + (sizeCell / 2), format);
            }
        }

        img.Save("FillGrid.png", ImageFormat.Png);
    }

    private void CreateAnswerFile(List<string> listDefinitionRight, List<string> listDefinitionDown)
    {
        var answerString = "По горизонтали: ";
        for (var i = 0; i < listDefinitionRight.Count; i++)
        {
            var newListWord = new List<string>(listDefinitionRight[i].Split(';'));
            var word = newListWord[1];
            answerString += newListWord[0] + ". " + word + ". ";
        }

        answerString += "\nПо вертикали: ";
        for (var i = 0; i < listDefinitionDown.Count; i++)
        {
            var newListWord = new List<string>(listDefinitionDown[i].Split(';'));
            var word = newListWord[1];
            answerString += newListWord[0] + ". " + word + ". ";
        }

        File.WriteAllText("Answer.txt", answerString);
    }

    private void CreateDefinitionFile(List<string> listDefinitionRight, List<string> listDefinitionDown, List<Dictionary> listDictionaries)
    {
        try
        {
            List<DictionaryWord> listWordsString = new();
            foreach (var dictionary in listDictionaries)
            {
                listWordsString.AddRange(dictionary.Words);
            }

            var definitionString = "По горизонтали: ";
            for (var i = 0; i < listDefinitionRight.Count; i++)
            {
                var newListWord = new List<string>(listDefinitionRight[i].Split(';'));
                var word1 = newListWord[1];
                foreach (var definition in listWordsString)
                {
                    var word2 = definition.Answers;
                    if (word1 != word2) continue;
                    var rnd = new Random();
                    var randomIndex = rnd.Next(0, definition.Definitions.Count);
                    definitionString += newListWord[0] + ". " + definition.Definitions[randomIndex] + ". ";
                    break;
                }
            }

            definitionString += "\nПо вертикали: ";
            for (var i = 0; i < listDefinitionDown.Count; i++)
            {
                var newListWord = new List<string>(listDefinitionDown[i].Split(';'));
                var word1 = newListWord[1];
                foreach (var definition in listWordsString)
                {
                    var word2 = definition.Answers;
                    if (word1 == word2)
                    {
                        var rnd = new Random();
                        var randomIndex = rnd.Next(0, definition.Definitions.Count);
                        definitionString += newListWord[0] + ". " + definition.Definitions[randomIndex] + ". ";
                        break;
                    }
                }
            }

            File.WriteAllText("Definition.txt", definitionString);
        }
        catch (Exception e)
        {
            _dialogService.ShowMessage("CreateDefinition\n" + e);
        }
    }
}

File: crossword_python/services/ScreenshotService.py
import random
from PIL import Image, ImageDraw, ImageFont
from crossword_python.models.Word import Word
from crossword_python.models.Dictionary import Dictionary
from crossword_python.viewmodels.CellViewModel import CellViewModel

class ScreenshotService:
    def __init__(self, dialog_service):
        self._dialog_service = dialog_service

    def export_crossword(self, list_words_grid: list[Word], list_dictionaries: list[Dictionary], cells: list[CellViewModel]):
        try:
            top_max_x = 99
            left_max_y = 99
            down_max_x = 0
            right_max_y = 0
            size_cell = 37.938105

            transparent_cells = [c for c in cells if c.background == "transparent"]
            if not transparent_cells:
                top_max_x = 1
                down_max_x = 0
            else:
                top_max_x = min(c.x for c in transparent_cells)
                left_max_y = min(c.y for c in transparent_cells)
                down_max_x = max(c.x for c in transparent_cells)
                right_max_y = max(c.y for c in transparent_cells)

            if down_max_x < top_max_x or right_max_y < left_max_y:
                self._dialog_service.show_message("Не найдено ячеек для создания скриншота. Убедитесь, что кроссворд сгенерирован.")
                return

            width = int((down_max_x - top_max_x + 1) * size_cell)
            height = int((right_max_y - left_max_y + 1) * size_cell)

            img_empty = Image.new('RGB', (width, height), color='white')
            draw_empty = ImageDraw.Draw(img_empty)
            horizontal_definitions = [ ]
            vertical_definitions = [ ]

            self._create_empty_grid(draw_empty, top_max_x, down_max_x, left_max_y, right_max_y, size_cell, horizontal_definitions, vertical_definitions, cells, list_words_grid)
            img_empty.save("EmptyGrid.png", format="PNG")

            img_fill = Image.new('RGB', (width, height), color='white')
            draw_fill = ImageDraw.Draw(img_fill)
            self._create_fill_grid(draw_fill, top_max_x, down_max_x, left_max_y, right_max_y, size_cell, cells)
            img_fill.save("FillGrid.png", format="PNG")

            self._create_answer_file(horizontal_definitions, vertical_definitions)
            self._create_definition_file(horizontal_definitions, vertical_definitions, list_dictionaries)
            self._dialog_service.show_message("Кроссворд сохранён")

        except Exception as e:
            self._dialog_service.show_message(f"Ошибка при создании скриншота:\n{str(e)}")

    def _create_empty_grid(self, draw: ImageDraw.ImageDraw, top_max_x: int, down_max_x: int, left_max_y: int, right_max_y: int, size_cell: float, horizontal_definitions: list[str], vertical_definitions: list[str], cells: list[CellViewModel], list_words_grid: list[Word]):
        try:
            font = ImageFont.truetype("arial.ttf", 12)
        except Exception:
            font = ImageFont.load_default()
        
        for cell in (c for c in cells if top_max_x <= c.x <= down_max_x and left_max_y <= c.y <= right_max_y):
            x0 = (cell.x - top_max_x) * size_cell
            y0 = (cell.y - left_max_y) * size_cell
            x1 = x0 + size_cell
            y1 = y0 + size_cell
            if cell.background == "transparent":
                draw.rectangle([x0, y0, x1, y1], fill="white", outline="black")
            else:
                draw.rectangle([x0, y0, x1, y1], fill="black")

        numbered_start_cells = {}
        number_counter = 1

        valid_words = [w for w in list_words_grid if len(w.cells) > 0]
        ordered_words = sorted(valid_words, key=lambda w: (w.cells[0].y, w.cells[0].x, 0 if w.right else 1))

        for word in ordered_words:
            start_cell = word.cells[0]
            cell_point = (start_cell.x, start_cell.y)

            if cell_point not in numbered_start_cells:
                word_number = number_counter
                number_counter += 1
                numbered_start_cells[cell_point] = word_number
                draw_x = (start_cell.x - top_max_x) * size_cell
                draw_y = (start_cell.y - left_max_y) * size_cell
                draw.text((draw_x + 2, draw_y + 2), str(word_number), fill="black", font=font)
            else:
                word_number = numbered_start_cells[cell_point]

            text = f"{word_number};{word.word_string}"
            if word.right:
                horizontal_definitions.append(text)
            else:
                vertical_definitions.append(text)

    def _create_fill_grid(self, draw: ImageDraw.ImageDraw, top_max_x: int, down_max_x: int, left_max_y: int, right_max_y: int, size_cell: float, cells: list[CellViewModel]):
        try:
            font = ImageFont.truetype("arial.ttf", 20)
        except Exception:
            font = ImageFont.load_default()
        
        for cell in (c for c in cells if top_max_x <= c.x <= down_max_x and left_max_y <= c.y <= right_max_y):
            x0 = (cell.x - top_max_x) * size_cell
            y0 = (cell.y - left_max_y) * size_cell
            x1 = x0 + size_cell
            y1 = y0 + size_cell

            if cell.background == "black":
                draw.rectangle([x0, y0, x1, y1], fill="black")
            else:
                draw.rectangle([x0, y0, x1, y1], fill="white", outline="black")
                content = cell.content.upper() if cell.content else ""
                if content:
                    # Базовое центрирование текста
                    draw.text((x0 + (size_cell / 3), y0 + (size_cell / 4)), content, fill="black", font=font)

    def _create_answer_file(self, horizontal_definitions: list[str], vertical_definitions: list[str]):
        answer_string = "По горизонтали: "
        for hd in horizontal_definitions:
            parts = hd.split(';')
            answer_string += f"{parts[0]}. {parts[1]}. "

        answer_string += "\nПо вертикали: "
        for vd in vertical_definitions:
            parts = vd.split(';')
            answer_string += f"{parts[0]}. {parts[1]}. "

        with open("Answer.txt", "w", encoding="utf-8") as f:
            f.write(answer_string)

    def _create_definition_file(self, horizontal_definitions: list[str], vertical_definitions: list[str], list_dictionaries: list[Dictionary]):
        list_words_string = [ ]
        for dictionary in list_dictionaries:
            list_words_string.extend(dictionary.words)

        definition_string = "По горизонтали: "
        for hd in horizontal_definitions:
            parts = hd.split(';')
            word1 = parts[1]
            for definition in list_words_string:
                if word1 == definition.answers and definition.definitions:
                    random_index = random.randint(0, len(definition.definitions) - 1)
                    definition_string += f"{parts[0]}. {definition.definitions[random_index]}. "
                    break

        definition_string += "\nПо вертикали: "
        for vd in vertical_definitions:
            parts = vd.split(';')
            word1 = parts[1]
            for definition in list_words_string:
                if word1 == definition.answers and definition.definitions:
                    random_index = random.randint(0, len(definition.definitions) - 1)
                    definition_string += f"{parts[0]}. {definition.definitions[random_index]}. "
                    break

        with open("Definition.txt", "w", encoding="utf-8") as f:
            f.write(definition_string)